### PR TITLE
[58011] Allow negative lag for predecessor/successor relations

### PR DESCRIPTION
**Pull request by cbliard**
# Ticket

https://community.openproject.org/wp/58011

# What are you trying to accomplish?

Allow negative lag for a predecessor/successor relation.

This makes it possible to have a follower starting the same day a predecessor ends, or at the same time as a predecessor starts.

## Screenshots

image

# What approach did you choose and why?

* Allow negative number and adapt calculation
* Add limits to lag: -2000 to +2000 days

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

### CODE REVIEW

File: app/models/relation.rb [hunk 1]
```diff
@@ -101,6 +101,9 @@ class Relation < ApplicationRecord
 
   ORDERED_TYPES = [*TYPES.keys, TYPE_PARENT, TYPE_CHILD].freeze
 
+  MIN_LAG = -2_000
+  MAX_LAG = 2_000
```
Comment (myabc) [inline app/models/relation.rb]: Very minor, but since these are additive inverses:

```suggestion
  MAX_LAG = 2_000
  MIN_LAG = -MAX_LAG
```

File: app/services/work_packages/shared/working_days.rb [hunk 2]
```diff
@@ -54,11 +54,20 @@ def lag(predecessor_date, successor_date)
...
         return nil unless date
 
-        date_after_lag = date + 1.day
         lag ||= 0
-        while lag > 0
-          lag -= 1 if working?(date_after_lag)
-          date_after_lag += 1
+        negative_lag = lag < 0
+
+        date_after_lag = date + 1.day
```
Comment (myabc) [inline app/services/work_packages/shared/working_days.rb]: I personally find the following reads a bit better:

```suggestion
        date_after_lag = date.next_day
```

or

```suggestion
        date_after_lag = date.succ
```

Comment (cbliard) [reply app/services/work_packages/shared/working_days.rb]: `next_day` will be perfect.

File: app/models/relation.rb (hunk 1 above)
Comment (cbliard) [reply app/models/relation.rb]: Good point.

**Comment by cbliard** [comment]
Thanks for the proposed improvements @myabc. I have integrated your suggestions and squashed the commits together.

I also had to split `add_lag` in 2 parts because it was becoming too complex to follow (`Metrics/PerceivedComplexity` was triggered).

File: spec/services/work_packages/shared/shared_examples_days.rb [hunk 3]
```diff
@@ -178,40 +178,56 @@
...
     end
 
     it "works with big lag value like 100_000" do
       # Ensure implementation is not recursive and won't fail with SystemStackError: stack level too deep
-      expect { subject.add_lag(sunday_2022_07_31, 100_000) }
+      expect { subject.with_lag(sunday_2022_07_31, 100_000) }
         .not_to raise_error
     end
 
     context "with weekend days (Saturday and Sunday)", :weekend_saturday_sunday do
```
Comment (myabc) [inline spec/services/work_packages/shared/shared_examples_days.rb]: I realise this is not part of the PR, but I find this context description ambiguous - I would expect _**with** weekend days_ to include Saturday and Sunday as working days,

Something more like?

- with Saturday and Sunday as weekend days
- with the weekend as non-working days
- without Saturday and Sunday

File: app/services/work_packages/shared/working_days.rb [hunk 4]
```diff
@@ -111,6 +110,26 @@ def assert_strictly_positive_duration(duration)
         raise ArgumentError, "duration must be strictly positive" if duration.is_a?(Integer) && duration <= 0
       end
 
+      def add_lag(date, lag)
+        date_with_lag = date.next_day
```
Comment (myabc) [inline app/services/work_packages/shared/working_days.rb:114]: I realise this is old behaviour (so apologies in advance if I misunderstand the desired behaviour), but in scenarios where the next day is a non-working day then we are returning that non-working day.

```ruby
context "with weekend days (Saturday and Sunday)", :weekend_saturday_sunday do
  friday_2025_05_30 = Date.new(2025, 5, 30)
  include_examples "it adds lag", date: friday_2025_05_30, lag: 0, expected_result: Date.new(2025, 5, 31) # result is Saturday - I would have expected Monday
  include_examples "it adds lag", date: friday_2025_05_30, lag: 1, expected_result: Date.new(2025, 6, 3) # result is next Tuesday - this makes sense

  include_examples "it adds lag", date: saturday_2022_07_30, lag: 0, expected_result: sunday_2022_07_31 # why Sunday, a non-working day?
...
end
```

File: app/services/work_packages/shared/working_days.rb [hunk 5]
```diff
@@ -111,6 +110,26 @@ def assert_strictly_positive_duration(duration)
...
+
+      def remove_lag(date, lag)
+        date_with_lag = date
+        while lag != 0
+          lag -= 1 if working?(date_with_lag)
+          next if lag == 0
+
+          date_with_lag -= 1
+        end
+        date_with_lag
```
Comment (myabc) [inline app/services/work_packages/shared/working_days.rb:130]: I had to solve a similar issue (working day calculations) at a previous job and accomplished it using an Enumerator. Feel free to use or not use this, but I personally find the functional style more readable vs. imperative.

```suggestion
        to_drop = [0, lag - 1].max
        Enumerator.produce(date, &:prev_day)
          .lazy
          .select { working?(it) }
          .drop(to_drop)
          .take(1)
          .first
```

I also tried to come up with a similar implementation for `add_tag` but struggled to get it working [with the above-mentioned behaviour](https://github.com/opf/openproject/pull/19033/files#r2109924606):

```ruby
      def add_lag(date, lag)
        to_drop = [0, lag].max
        Enumerator.produce(date, &:next_day)
          .lazy
          .drop(1)
          .select { working?(it) }
          .drop(to_drop)
          .take(1)
          .first
      end
```

File: spec/models/relation_spec.rb [hunk 6]
```diff
@@ -38,17 +38,24 @@
...
 
-    relation.lag = 2_147_483_648
+    relation.lag = -2_001
     expect(relation).not_to be_valid
-    expect(relation.errors[:lag]).to include(I18n.t(:"activerecord.errors.models.relation.attributes.lag.less_than_or_equal_to"))
-
-    relation.lag = 1_000
-    expect(relation).to be_valid
+    expect(relation.errors[:lag])
+      .to eq(["must be greater than or equal to #{Relation::MIN_LAG}."])
```
Comment (myabc) [inline spec/models/relation_spec.rb]: I personally find [Shoulda Matchers](https://matchers.shoulda.io/) quite helpful when testing AR validations. You could use the [`validate_numericality_of`](https://matchers.shoulda.io/docs/v6.5.0/Shoulda/Matchers/ActiveModel.html#validate_numericality_of-instance_method) or [`allow_value`](https://matchers.shoulda.io/docs/v6.5.0/Shoulda/Matchers/ActiveModel.html#allow_value-instance_method) matchers.

```suggestion
   expect(relation).to validate_numericality_of(:lag)
     .is_greater_than_or_equal_to(-2_000)
     .is_less_than_or_equal_to(2_000)
     .allow_nil
```

File: .rubocop.yml [hunk 7]
```diff
@@ -204,6 +204,12 @@ RSpec/DescribeClass:
     - 'spec/features/**/*.rb'
     - 'modules/*/spec/features/**/*.rb'
 
+RSpec/IncludeExamples:
+  Exclude:
+    - 'spec/services/work_packages/shared/shared_examples_days.rb'
+    - 'spec/services/work_packages/shared/working_days_spec.rb'
+    - 'spec/services/work_packages/shared/all_days_spec.rb'
```
Comment (myabc) [inline .rubocop.yml]: I presume you're waiting for #19040 to be merged before removing?

File: app/services/work_packages/shared/working_days.rb (hunk 4 above)
Comment (cbliard) [reply app/services/work_packages/shared/working_days.rb:114]: Yes, it's because if the successor ignores non-working days, you want it to start on Saturday. 

That's this scenario: 
* predecessor ends on Thursday
* predecessor-successor relation has lag = 1 day (1 working day between predecessor and successor)
* successor has ignore non-working days = true
* => the earliest the successor can start is on Saturday, not on Monday

File: app/services/work_packages/shared/working_days.rb (hunk 5 above)
Comment (cbliard) [reply app/services/work_packages/shared/working_days.rb:130]: Maybe I'm not used to it, but I actually find it harder to read. It's ok until `select { working?(it) }`, but then `drop([0, lag - 1].max)`, and `take(1)` and `first`. I don't know, maybe not my way of seeing the problem.

And for debugging, chained calls are a chore too. Putting a `binding.irb` inside the `while` loop made it easier to debug in my early implementations. It's possible to put some `.tap { binding.irb }` here and there, but I find it less intuitive.

Sorry, I'll keep it like this.

File: spec/models/relation_spec.rb (hunk 6 above)
Comment (cbliard) [reply spec/models/relation_spec.rb]: True, but I always felt a bit hesitant with these shoulda matchers, because they really just seem to duplicate what's in the model, just written another way:

```ruby
   expect(relation).to validate_numericality_of(:lag)
     .is_greater_than_or_equal_to(-2_000)
     .is_less_than_or_equal_to(2_000)
     .allow_nil
```
vs
```ruby
  validates :lag, numericality: {
    allow_nil: true,
    less_than_or_equal_to: 2_000,
    greater_than_or_equal_to: -2_000
  }
```

That's the same code.
The test only ensures that this validation is never deleted, and creates a maintenance burden whenever the values in the validator are changed.

To be honest I was hesitant to just delete that test. It was doing more before because the i18n was different, so it kinda made sense, but now I don't feel it's that useful.

Nevertheless, if we are to keep that test, then using should matchers is more expressive. I'll use your proposition.

File: .rubocop.yml (hunk 7 above)
Comment (cbliard) [reply .rubocop.yml]: ~No, quite the oppposite~. Well I wanted to merge that one first because the tests were passing. But with your review I'll have to amend and force push it again anyway, so yes that's probably the right thing to do.

File: spec/services/work_packages/shared/shared_examples_days.rb (hunk 3 above)
Comment (cbliard) [reply spec/services/work_packages/shared/shared_examples_days.rb]: Good point. Updated to "with Saturday and Sunday being non-working days" as "weekend" has different meanings depending on your country / culture.

File: spec/models/relation_spec.rb (hunk 6 above)
Comment (cbliard) [reply spec/models/relation_spec.rb]: The failure message with shoulda_matchers is particularly nice:

```
  1) Relation validates lag numericality
     Failure/Error:
       expect(relation).to validate_numericality_of(:lag)
         .is_greater_than(Relation::MIN_LAG)
         .is_less_than_or_equal_to(Relation::MAX_LAG)
         .allow_nil

       Expected Relation to validate that :lag looks like a number greater than
       -2000 and less than or equal to 2000 as long as it is not nil, but this
       could not be proved.
         In checking that Relation disallows :lag from being a number that is
         not greater than -2000, after setting :lag to ‹-2000›, the matcher
         expected the Relation to be invalid, but it was valid instead.
```

Comment (cbliard) [reply spec/models/relation_spec.rb]: And actually this test will be useful when we'll restrict the lag property to relations of type follows/precedes. It makes no sense to allow setting lag for a relation of type relates_to (for instance). Jens created a bug ticket a while ago about that.

**Comment by cbliard** [comment]
Thanks for the code review @myabc. I do value code reviews, and I value behavior reviews even more as that's what matters most in the end.

Did you had a chance to test the feature and use negative lag in the app? Is it working as expected? Did you find some edge cases?

File: app/services/work_packages/shared/working_days.rb (hunk 5 above)
Comment (myabc) [reply app/services/work_packages/shared/working_days.rb:130]: No worries. The original is also not not clear.

File: app/services/work_packages/shared/working_days.rb (hunk 4 above)
Comment (myabc) [reply app/services/work_packages/shared/working_days.rb:114]: Thanks for clarifying the expected behaviour!

**Comment by myabc** [comment]
> Did you had a chance to test the feature and use negative lag in the app? Is it working as expected? Did you find some edge cases?

Yes, I did checkout the branch and test. It works as expected (thanks again for [clarifying the expected behaviour with regards to non-working days](https://github.com/opf/openproject/pull/19033#discussion_r2111099190))

**Review by myabc — APPROVED** [review]
Looks good! 👍🏻 (CI failure does not appear to be related)
